libs/cardano-ledger-conformance/src/Test/Cardano/Ledger/Conformance.hs

```diff
@@ -418,3 +513,142 @@ instance
+          <> show (all (\txout -> serSize (txout ^. valueTxOutL) <= maxValSize) txouts)
+      , " ∀[ (a , _) ∈ range txouts ] Sum.All (const ⊤) (λ a → a .BootstrapAddr.attrsSize ≤ 64) a\n  "
+          -- I assume we never create bootstrap addresses
+          <> show (all (\txout -> case txout ^. addrTxOutL of AddrBootstrap _ -> False; _ -> True) txouts)
+      ]
+  logEntry $
+    unlines
+      [ "Implementation:"
+      , "  consumed:\t" ++ show (coin consumedVal)
+      , "    ins:\t\t" ++ show (coinBalance (UTxO (Map.restrictKeys (unUTxO utxo) (txb ^. inputsTxBodyL))))
+      , "  produced:\t" ++ show (coin producedVal)
+      , "    outs:\t" ++ show (coin $ sumAllValue txouts)
+      , "    fee:\t\t" ++ show txfee
+      ]
+  lastTick <- gets $ view impLastTickL
+  pParams <- getsNES $ nesEsL . curPParamsEpochStateL
+  certState <- getsNES $ nesEsL . esLStateL . lsCertStateL
+  let
+    utxoEnv =
+      UtxoEnv
+        { ueSlot = lastTick
+        , uePParams = pParams
+        , ueGenDelegs = certState ^. certDStateL . dsGenDelegsL
+        , ueCertState = certState
+        }
+  agdaUtxoEnv <- case toSpecRep utxoEnv of
+    Right x -> pure x
+    Left e -> fail $ T.unpack e
+  let
+    newAgdaUtxoState = Agda.utxoStep agdaUtxoEnv agdaUtxoState agdaTx
+  modify $ impDoFixupL .~ False
+  submitRes <- trySubmitTx tx
+  modify $ impDoFixupL .~ doFixup
+  newUtxoState <- getsNES $ nesEsL . esLStateL . lsUTxOStateL
+  let
+    finalAgdaState =
+      toExpr . specToTestRep @(UTxOState Conway) <$> newAgdaUtxoState
+    finalImplState = case submitRes of
+      Right _ -> case toTestRep newUtxoState of
+        Right x -> Just $ toExpr x
+        Left _ -> Nothing
+      Left e ->
+        error $
+          "Failed to submit transaction on the implementation side:\n"
+            <> show e
+  unless (finalAgdaState == finalImplState) $ do
+    logEntry $ diffExprCompact finalAgdaState finalImplState
+    fail "The final states of spec and implementation do not match"
+  pure submitRes
+
+spec :: HasCallStack => Spec
+spec = describe "Conway conformance tests" . withImpState $ do
+  it "Simple tx" $ do
+    modify $ \x ->
+      x
+        & impLastTickL .~ 100
+    do
+      certState <- getsNES $ nesEsL . esLStateL . lsCertStateL
+      govState <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosGovStateL
+      totalObligation certState govState `shouldBe` zero
+    do
+      deposited <- getsNES $ nesEsL . esLStateL . lsUTxOStateL . utxosDepositedL
+      deposited `shouldBe` zero
+    res <- trySubmitTxConform $ mkBasicTx mkBasicTxBody
+    case res of
+      Right _ -> pure ()
+      Left e -> error $ show e
+  it "Constrained generator" $ do
+    let
+      us =
+        def
+          { usMinCerts = 0
+          , usMaxCerts = 0
+          , usDatumFreq = 0
+          , usGenerateWithdrawals = False
+          , usNumPreUtxo = 3
+          , usNumColUtxo = 3
+          }
+    modify $ \x ->
```

Comment:
I was sure I submitted this comment yesterday, but maybe I forgot to click the "Add" button:
I am pretty sure you can use `prop` with ImpTest if you want to have access to quickcheck generators:
```suggestion
  prop "Constrained generator" $
    let
      us =
        def
          { usMinCerts = 0
          , usMaxCerts = 0
          , usDatumFreq = 0
          , usGenerateWithdrawals = False
          , usNumPreUtxo = 3
          , usNumColUtxo = 3
          }
    in forAll (genTxAndNewEpoch us $ reify @Conway) $ \(nes, tx, _env) -> do          
      modify $ \x ->
```